Crash in agent.go

Hit this panic which crashed docker daemon on `docker swarm init` while testing some swarm changes (swarm code off of 036a4a1e934bd1bbb35c3ec7f85dea2ba6d4e336). Although my changes do not touch this code, they seem to cause a session reset which then sometimes leads to the panic.

I do not have enough context to say whether it is a legit scenario, but I think somebody with more knowledge about the code flow should check if worth to check ` if backoff <= 0 {` before passing it to `rand.Int63n()`


```
DEBU[0020] (*session).heartbeat                          module=agent session.id=cuwhctdjtieynaeydfwl0fxh9
DEBU[0020] (*session).listen                             module=agent session.id=cuwhctdjtieynaeydfwl0fxh9
DEBU[0020] (*session).watch                              module=agent session.id=cuwhctdjtieynaeydfwl0fxh9
DEBU[0020] (*session).watchAttachments                   module=agent session.id=cuwhctdjtieynaeydfwl0fxh9
DEBU[0020]                                               method=(*Dispatcher).Tasks node.id=9m4uvbpm6tkjo5mtoh7s1d90q node.session=cuwhctdjtieynaeydfwl0fxh9

Registering for attachments
DEBU[0020] MYCODE                                        method=(*Dispatcher).ExecutorAttachments node.id=9m4uvbpm6tkjo5mtoh7s1d90q node.session=cuwhctdjtieynaeydfwl0fxh9
ERRO[0020] agent: session failed                         error=rpc error: code = 2 desc = invalid find argument type module=agent
DEBU[0020] agent: registered                             module=agent
DEBU[0020] agent: rebuild session                        module=agent
DEBU[0020] (*Agent).run exited                           module=agent
panic: invalid argument to Int63n

goroutine 123 [running]:
panic(0x1779a80, 0xc820a6d580)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
math/rand.(*Rand).Int63n(0xc82000e2f0, 0x0, 0x1)
	/usr/local/go/src/math/rand/rand.go:64 +0x79
math/rand.Int63n(0x0, 0x1f3a9c0)
	/usr/local/go/src/math/rand/rand.go:203 +0x2d
github.com/docker/swarmkit/agent.(*Agent).run(0xc820a850a0, 0x7fcbb7656f30, 0xc8208d00f0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/swarmkit/agent/agent.go:206 +0x1523
created by github.com/docker/swarmkit/agent.(*Agent).Start
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/swarmkit/agent/agent.go:81 +0x37d
```

--- a/agent/agent.go
+++ b/agent/agent.go
@@ -195,6 +195,8 @@ func (a *Agent) run(ctx context.Context) {
 				log.G(ctx).WithError(err).Error("agent: closing session failed")
 			}
 			sessionq = nil
+			// if we're here before <-registered, do nothing for that event
+			registered = nil
 		case <-session.closed:
 			log.G(ctx).Debugf("agent: rebuild session")
 

